tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py

```diff
@@ -47,24 +47,71 @@ def from_msg_msg(msg):
 tf2_ros.ConvertRegistration().add_from_msg(PointCloud2, from_msg_msg)
 
 
-def transform_to_kdl(t):
-    return PyKDL.Frame(PyKDL.Rotation.Quaternion(
-                            t.transform.rotation.x, t.transform.rotation.y,
-                            t.transform.rotation.z, t.transform.rotation.w),
-                       PyKDL.Vector(t.transform.translation.x,
-                                    t.transform.translation.y,
-                                    t.transform.translation.z))
+def transform_points(point_cloud: np.ndarray, transform: Transform) -> np.ndarray:
+    """
+    Transforms a bulk of points from an numpy array using a provided `Transform`.
+
+    :param point_cloud: nx3 Array of points where n is the number of points
+    :param transform: TF2 transform used for the transformation
+    :returns: Array with the same shape as the input array, but with the transformation applied
+    """
+    # Build affine transformation
+    transform_translation = np.array([
+        transform.translation.x,
+        transform.translation.y,
+        transform.translation.z
+    ])
+    transform_rotation_matrix = _get_mat_from_quat(
+        np.array([
+            transform.rotation.w,
+            transform.rotation.x,
+            transform.rotation.y,
+            transform.rotation.z
+        ]))
+
+    # "Batched" matmul meaning a matmul for each point
+    # First we offset all points by the translation part followed by a rotation using the rotation matrix
+    return np.einsum("ij, pj -> pi", transform_rotation_matrix, point_cloud + transform_translation)
+
+
 
+def do_transform_cloud(cloud: PointCloud2, transform: Transform) -> PointCloud2:
+    """
+    Applies a `Transform` on a `PointCloud2`.
 
-# PointStamped
-def do_transform_cloud(cloud, transform):
-    t_kdl = transform_to_kdl(transform)
-    points_out = []
-    for p_in in read_points(cloud):
-        p_out = t_kdl * PyKDL.Vector(p_in[0], p_in[1], p_in[2])
-        points_out.append(p_out)
-    res = create_cloud(transform.header, cloud.fields, points_out)
-    return res
+    :param cloud: The point cloud that should be transformed
+    :param transform: The transform which will applied to the point cloud
+    :returns: The transformed point cloud
+    """
+    points = read_points(cloud)
+    points_out = transform_points(points, transform)
+    return create_cloud(transform.header, cloud.fields, points_out)
 
 
 tf2_ros.TransformRegistration().add(PointCloud2, do_transform_cloud)
+
+
+def _get_mat_from_quat(quaternion: np.ndarray):
```

Comment:
i.e.
```suggestion
def _get_mat_from_quat(quaternion: npt.ArrayLike) -> npt.ArrayLike:
```

Reply:
`npt.ArrayLike` is defined as everything that can be cast into a NumPy array using e.g. the `np.asarray()` method. This included scalars as well as lists, NumPy arrays, and so on. So we could extend this interface to include all of these, but we would also need to add an `np.asarray()` call for the input. I think it is okay to type it as a `np.ndarray`, because it is a private function which is only used with NumPy arrays. 

Also the output is strictly a `np.ndarray` and not a `npt.ArrayLike`

See: https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike